### PR TITLE
Add image support to the text editor

### PR DESCRIPTION
**Pull request by kesselb**
Fixes https://github.com/nextcloud/mail/issues/2733

### Changes

- TextEditor: Rework the data flow between Composer and TextEditor component.
- TextEditor: Add editorExecute to call ckeditor plugins externally.
- Signature: Schema is defined as $container now. Should allow user to edit the signature in message composer.
- Signature: Run signature command only on alias change / compose new message.
- Backend: Add parser for DataUri.
- Backend: Load, modify and save HTML messages via DomDocument.
- Backend: Component to build a multipart/related message

### To do

- [ ] Testing with Mailvelope
- [x] Convert base64 image to attachment
- [x] Evaluate isHtml flag for drafts when loading a message with a different format then account default (rich vs plain)
- [x] Set editor mode only for new messages
- [x] Check DomDocument
- [x] Debug: `mail: setting TextEditor contents to <undefined>`
- [x] Backend: Create a multipart/related message for html, text and inline images
- [ ] Tests (onAliasChange)
- [x] Align list of allowed mime types (ckeditor / html purifier)

### To discuss

- Have account A (plain text) and account B (rich text). What should happen switching from rich text to plain text? 
- ~Use the new HTML2Text dependency for html-to-text conversion instead of horde filter?~
- ~Caption/Alt for Image? A possible use case is the attachment name~
- ~Add a warning for images > 128kb? (big base64 encoded images make the editor laggy and slow down sending)~
- ~https://github.com/nextcloud/mail/pull/6935/commits/ca38fb75d175ec02ab7248bddedd792e368c830b add debounce for onInput event~

### Test cases

**1. Open draft**

- New message
- Insert image
- Wait for draft saved
- Close editor 
- Open drafts folder
- Open message again

Expected result: Image is there. It's possible to add further images, delete existing images and write more text.

**2. Write message**

- New message
- Insert image
- Send unencrypted

Expected result: Message is in the outbox. Image is there. A `multipart/related` email is sent. K-9, Thunderbird, Roundcube show the message properly.

### To rebase the branch with main

- Run interactive rebase
- Drop commit "Add ckeditor upload and image as dependency"
- `npm install @ckeditor/ckeditor5-image @ckeditor/ckeditor5-upload @ckeditor/ckeditor5-vue2@^3.0.1`
- `git add package.json package-lock.json`
- `git commit --signoff -m "Add ckeditor upload and image as dependency"`

### CODE REVIEW

**Comment by miaulalala** [comment]
oof

**Comment by ChristophWurst** [comment]
Nitpick: the image tool is very hidden 
Bildschirmfoto vom 2022-08-26 16-12-40

**Comment by ChristophWurst** [comment]
- [ ] BUG weird behavior with image in signature but plain editing as default

1) Set the account to plain text emails
2) Add an image to your signature
3) Compose a message
4) Switch to rich text

Expected: the image appears
Actual: the image does not show

**Comment by ChristophWurst** [comment]
:heavy_check_mark: Successfully tested embedding an image into a new HTML message
:heavy_check_mark: Successfully tested embedding an image into my signature and sending an HTML message

**Comment by ChristophWurst** [comment]
> Have account A (plain text) and account B (rich text). What should happen switching from rich text to plain text?

We answered that, right? If there is content, the editing mode won't switch with the selection of another account. If the body is empty we switch modes according to account defaults.

**Comment by ChristophWurst** [comment]
> Add a warning for images > 128kb? (big base64 encoded images make the editor laggy and slow down sending)

Definitely. Does CKEditor have an option for this or allow us to hook into the selection?

**Comment by ChristophWurst** [comment]
Adjusted the title so it can be found in the browser history

File: src/components/TextEditor.vue [hunk 1]
```diff
@@ -250,3 +250,13 @@ export default {
 	cursor: text;
 }
 </style>
+
+<style>
+/*
+Overwrite the default z-index for CKEditor
+https://github.com/ckeditor/ckeditor5/issues/1142
```
Comment (ChristophWurst) [inline src/components/TextEditor.vue:257]: do you know whly this only happens here but not on `main`?

File: src/components/TextEditor.vue [hunk 2]
```diff
@@ -187,29 +216,30 @@ export default {
...
-
 			this.bus.$on('append-to-body-at-cursor', this.appendToBodyAtCursor)
-			this.bus.$on('insert-signature', this.onInsertSignature)
+			this.$emit('ready', editor)
 		},
-		onInput() {
-			logger.debug(`TextEditor input changed to <${this.text}>`)
-			this.$emit('input', this.text)
+		onEditorInput(text) {
+			text = this.sanitizeValue(text)
```
Comment (kesselb) [inline src/components/TextEditor.vue:223]: Inconvenient to call DomPurify here again ...

**Comment by kesselb** [comment]
> https://github.com/nextcloud/mail/pull/6935/commits/ca38fb75d175ec02ab7248bddedd792e368c830b add debounce for onInput event

Resolved via https://github.com/nextcloud/mail/pull/6935/commits/ea90bc1f4c668592e9d7d0ee1dfdf787077cd60b

File: src/components/Composer.vue [hunk 3]
```diff
@@ -990,11 +989,16 @@ export default {
 			}
 			return new Date(timestamp).toLocaleString(getCanonicalLocale(), options)
 		},
-		onAliasChange(alias) {
+		async onAliasChange(alias) {
```
[the remark this answers is not in the document]
Comment (kesselb) [reply src/components/Composer.vue]: Resolved via https://github.com/nextcloud/mail/pull/6935/commits/b3ad2758003eb776dbe4a4f29cf9e98f010ad61a

File: src/components/TextEditor.vue (hunk 2 above)
Comment (kesselb) [reply src/components/TextEditor.vue]: Resolved via https://github.com/nextcloud/mail/commit/ea90bc1f4c668592e9d7d0ee1dfdf787077cd60b

**Review by ChristophWurst — APPROVED** [review]
Tested & works

**Pull request merged**